src/librawspeed/common/Common.h

```diff
@@ -130,21 +130,39 @@ inline Endianness getHostEndianness() {
 # define BSWAP64(A) __builtin_bswap64(A)
 #endif
 
-template<typename T> inline T getByteSwapped(const void* data, bool bswap)
+inline short16 getByteSwapped(short16 v) { return BSWAP16(v); }
+inline ushort16 getByteSwapped(ushort16 v) { return BSWAP16(v); }
+inline int32 getByteSwapped(int32 v) { return BSWAP32(v); }
+inline uint32 getByteSwapped(uint32 v) { return BSWAP32(v); }
+inline uint64 getByteSwapped(uint64 v) { return BSWAP64(v); }
+
+// the float/double versions use two memcpy which guarantee strict aliasing
+// and are compiled into the same assembly as the popular union trick.
+inline float getByteSwapped(float f)
+{
+  uint32 i;
+  memcpy(&i, &f, sizeof(i));
+  i = BSWAP32(i);
+  memcpy(&f, &i, sizeof(i));
+  return f;
+}
+inline double getByteSwapped(double d)
+{
+  uint64 i;
+  memcpy(&i, &d, sizeof(i));
+  i = BSWAP64(i);
+  memcpy(&d, &i, sizeof(i));
+  return d;
+}
+
+template <typename T> inline T getByteSwapped(const void* data, bool bswap)
 {
   T ret;
   // all interesting compilers optimize this memcpy into a single move
-  // this is the most effective way to load some bytes without running into alignmen issues
+  // this is the most effective way to load some bytes without running into
+  // alignment or aliasing issues
   memcpy(&ret, data, sizeof(T));
-  if (bswap) {
-    switch(sizeof(T)) {
-    case 1: break;
-    case 2: ret = BSWAP16(ret); break;
-    case 4: ret = BSWAP32(ret); break;
-    case 8: ret = BSWAP64(ret); break;
-    }
-  }
-  return ret;
+  return bswap ? getByteSwapped(ret) : ret;
```

Comment:
> and it should be much faster.

It is indeed. thanks.

> Also, you do know that you can do ctest -j9?

No I did not. It is not that I really cared about the 30 seconds test run. I mainly wanted to point out that your testing methodology could be improved. I think a white-box testing approach is the right tool for this kind of problem and the complete set of those byte swap functions has on the order of 10 code paths (did not count). What I am ultimately trying to say is: in my personal opinion you could have spent your time and effort in this particular case to produce something with more value for the project than those bash-scripted black-box test numbers.

> Just putting all the cards on the table here.

Thanks for that. Appreciated. Just my 2 cents, though: as a respectful maintainer, maybe the good of the project could use a little bump in your priority list ;).

> But I need to be motivated to review&merge things.

Agreed. I did this hacking marathon for fun and I don't expect others to do stuff they don't like. But reviewing PRs is part of the responsibility of the role _you gave yourself_. If you don't like that just do something else. If that means not looking into code from others anymore, so be it. But if your sense of 'ownership' for the code (most of which you did not write) means forking this project at this point in time for this reason would be just sad (IMHO).

Anyway, I said I was about done with my enthusiasm working on this code and that is still true. I simply wanted to leave something 'consistent' and therefore did the Tiff interface, FILEMAP_MARGIN and DngOpcodes pr. The `Hint` class was just a little thing that seemed so obvious, I could not resist. Same goes for the Coverity fixes. I was curious if they would make sense now that you changed the compiler and while going through them one by one, why not fix them right away...

The only thing left in my patch queue at the moment is the introduction of some 'size/offset' type into the BitStream and TiffEntry interfaces. I was still waiting on some final response from you on the naming question in #76 (which really is only about "which string is nicer to look at: 'u16' or 'uint16_t'?")

TLDR; I'll let you hack on in peace soon ;).